### PR TITLE
[feature-wip] (array-type) add the array_sort function

### LINKED ISSUE #10052: [Feature] Add lots of Array Functions
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

We have supported `array` data type recently, all pr(s) about `array` please refer to https://github.com/apache/incubator-doris/issues/7570 .

Now we need to add lots of array functions to make `array` more powerful.
If anybody is interesting to `array`, please let me know.

Array Functions DONE:
- size/cardinality
- array_position
- array_contains
- element_at

Array Functions DOING:
- array_sum
- array_avg
- array_min
- array_max
- array_product
- array_distinct
- array_remove
- arrays_overlap

Array Functions TODO:
- array_union
- array_intersect
- array_except
- reverse
- concat_ws
- array_join
- array_sort
- array_slice

### Use case

_No response_

### Related issues

https://github.com/apache/incubator-doris/issues/7570

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by carlvinhust2012**
# Proposed changes
Issue Number: close #10052
this pr is used to add the array_sort function for array-type.
we can run this function in sql as follow:
MySQL [(none)]> set enable_array_type=true;

MySQL [(none)]> set enable_vectorized_engine=true;

MySQL [example_db]> select id, c_array, array_sort(c_array) from example_db.array_test;
+------+-----------------------------+-----------------------------+
| id   | c_array                     | array_sort(`c_array`)       |
+------+-----------------------------+-----------------------------+
|    1 | [1, 2, 3, 4, 5]             | [1, 2, 3, 4, 5]             |
|    2 | [6, 7, 8]                   | [6, 7, 8]                   |
|    3 | []                          | []                          |
|    4 | NULL                        | NULL                        |
|    5 | [1, 2, 3, 4, 5, 4, 3, 2, 1] | [1, 1, 2, 2, 3, 3, 4, 4, 5] |
|    6 | [1, 2, 3, NULL]             | [NULL, 1, 2, 3]             |
|    7 | [1, 2, 3, NULL, NULL]       | [NULL, NULL, 1, 2, 3]       |
|    8 | [1, 1, 2, NULL, NULL]       | [NULL, NULL, 1, 1, 2]       |
|    9 | [1, NULL, 1, 2, NULL, NULL] | [NULL, NULL, NULL, 1, 1, 2] |
+------+-----------------------------+-----------------------------+
9 rows in set (0.010 sec)

MySQL [example_db]> select id, c_array, array_sort(c_array) from example_db.array_test01;
+------+--------------------------+--------------------------+
| id   | c_array                  | array_sort(`c_array`)    |
+------+--------------------------+--------------------------+
|    1 | [a, b, c, d, e]          | [a, b, c, d, e]          |
|    2 | [f, g, h]                | [f, g, h]                |
|    3 | []                       | []                       |
|    3 | [NULL]                   | [NULL]                   |
|    5 | [a, b, c, d, e, a, b, c] | [a, a, b, b, c, c, d, e] |
|    6 | NULL                     | NULL                     |
|    7 | [a, b, NULL]             | [NULL, a, b]             |
|    8 | [a, b, NULL, NULL]       | [NULL, NULL, a, b]       |
+------+--------------------------+--------------------------+
8 rows in set (0.017 sec)

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No)
3. Has document been added or modified: (No)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/array/function_array_sort.h [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+        }
+        return true;
+    }
+
+    bool _execute_by_type(const IColumn& src_column, const ColumnArray::Offsets& src_offsets,
+                          IColumn& dest_column, ColumnArray::Offsets& dest_offsets,
+                          const NullMapType* src_null_map, NullMapType* dest_null_map,
+                          DataTypePtr& nested_type) {
+        bool res = false;
+        WhichDataType which(remove_nullable(nested_type)->get_type_id());
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_sort.h:246]:
```suggestion
        WhichDataType which(remove_nullable(nested_type));
```

File: be/src/vec/functions/array/function_array_sort.h [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+                    (*dest_null_map).push_back(true);
+                    dest_datas.push_back(NestType());
+                }
+            }
+
+            _sort_by_permutation<ColumnType>(prev_src_offset, curr_src_offset, src_data_concrete,
+                                             src_column, src_null_map, permutation);
+
+            // insert non-null element after sort by permutation
+            for (ColumnArray::Offset j = prev_src_offset; j < curr_src_offset; ++j) {
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_sort.h:164]: it looks like, the two for-loop can merge to one.
// filter and insert null element first
// insert non-null element after sort by permutation

File: be/src/vec/functions/array/function_array_sort.h [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+            return false;
+        }
+        const PaddedPODArray<NestType>& src_datas = src_data_concrete->get_data();
+
+        ColumnType& dest_data_concrete = reinterpret_cast<ColumnType&>(dest_column);
+        PaddedPODArray<NestType>& dest_datas = dest_data_concrete.get_data();
+
+        ColumnArray::Offset prev_src_offset = 0;
+        IColumn::Permutation permutation(src_column.size());
+        for (size_t i = 0; i < src_column.size(); ++i) {
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_sort.h:146]: we should init permutation for each row?

File: be/src/vec/functions/array/function_array_sort.h (hunk 2 above)
Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_sort.h:164]: if the array is [1, null, 2, null, 3] and just use one loop, the dest_datas will push_back [1, null, 2, null, 3].

File: be/src/vec/functions/array/function_array_sort.h (hunk 3 above)
Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_sort.h:146]: the src_column.size() is the total number of the array elements.

File: be/src/vec/functions/array/function_array_sort.h [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+            permutation[i] = i;
+        }
+
+        for (auto curr_src_offset : src_offsets) {
+            // filter and insert null element first
+            for (ColumnArray::Offset j = prev_src_offset; j < curr_src_offset; ++j) {
+                if (src_null_map && (*src_null_map)[j]) {
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_sort.h:153]: put src_null_map outside of for loop

File: be/src/vec/functions/array/function_array_sort.h [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+            // insert non-null element after sort by permutation
+            for (ColumnArray::Offset j = prev_src_offset; j < curr_src_offset; ++j) {
+                if (src_null_map && (*src_null_map)[j]) {
+                    continue;
+                }
+
+                StringRef src_str_ref = src_data_concrete->get_data_at(permutation[j]);
+                // copy the src data to column_string_chars
+                const size_t old_size = column_string_chars.size();
+                const size_t new_size = old_size + src_str_ref.size + 1;
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_sort.h:222]: no need to append zero char.

File: be/src/vec/functions/array/function_array_sort.h (hunk 4 above)
Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_sort.h:153]: I have re-check here. Even if it has been modified, the improvement is limited.

File: be/src/vec/functions/array/function_array_sort.h (hunk 2 above)
Comment (cambyzju) [reply be/src/vec/functions/array/function_array_sort.h:164]: got it, array_sort will put all nulls in front of array.

**Review by xy720 — DISMISSED** [review]
LGTM

**Comment by xy720** [comment]
Please rebase master.

**Comment by carlvinhust2012** [comment]
> Please rebase master.

ok

**Comment by cambyzju** [comment]
LGTM

**Review by xy720 — APPROVED** [review]
LGTM

**Pull request merged**